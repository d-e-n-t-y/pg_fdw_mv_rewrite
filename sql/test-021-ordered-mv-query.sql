-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key;

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by 1;

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc;

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key desc;

-- Anticipate rewrite
set mv_rewrite.log_match_progress = true;
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc nulls last;

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc nulls last;

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key, count (value);

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by count (value), key;
