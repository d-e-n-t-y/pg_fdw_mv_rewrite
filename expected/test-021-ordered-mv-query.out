-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key;
              QUERY PLAN              
--------------------------------------
 Custom Scan (MVRewriteScan)
   Output: key, (count(value))
   Rewritten: scan of public.test_mv5
   ->  Seq Scan on public.test_mv5
         Output: key, count
(5 rows)

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by 1;
              QUERY PLAN              
--------------------------------------
 Custom Scan (MVRewriteScan)
   Output: key, (count(value))
   Rewritten: scan of public.test_mv5
   ->  Seq Scan on public.test_mv5
         Output: key, count
(5 rows)

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc;
              QUERY PLAN              
--------------------------------------
 Custom Scan (MVRewriteScan)
   Output: key, (count(value))
   Rewritten: scan of public.test_mv5
   ->  Seq Scan on public.test_mv5
         Output: key, count
(5 rows)

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key desc;
                 QUERY PLAN                 
--------------------------------------------
 Sort
   Output: key, (count(value))
   Sort Key: test.key DESC
   ->  Custom Scan (MVRewriteScan)
         Output: key, (count(value))
         Rewritten: scan of public.test_mv1
         ->  Seq Scan on public.test_mv1
               Output: key, count
(8 rows)

-- Anticipate rewrite
set mv_rewrite.log_match_progress = true;
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc nulls last;
INFO:  evaluating MV: public.test_mv1
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
              QUERY PLAN              
--------------------------------------
 Custom Scan (MVRewriteScan)
   Output: key, (count(value))
   Rewritten: scan of public.test_mv5
   ->  Seq Scan on public.test_mv5
         Output: key, count
(5 rows)

-- Anticipate rewrite
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key asc nulls last;
INFO:  evaluating MV: public.test_mv1
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
              QUERY PLAN              
--------------------------------------
 Custom Scan (MVRewriteScan)
   Output: key, (count(value))
   Rewritten: scan of public.test_mv5
   ->  Seq Scan on public.test_mv5
         Output: key, count
(5 rows)

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by key, count (value);
INFO:  evaluating MV: public.test_mv1
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has fewer sort keys.
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has fewer sort keys.
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  no candidate MVs for query involving {public.test_mv5}.
                 QUERY PLAN                 
--------------------------------------------
 Sort
   Output: key, (count(value))
   Sort Key: test.key, (count(test.value))
   ->  Custom Scan (MVRewriteScan)
         Output: key, (count(value))
         Rewritten: scan of public.test_mv1
         ->  Seq Scan on public.test_mv1
               Output: key, count
(8 rows)

-- Anticipate rewrite, but not using _mv5
explain (verbose, costs off) select key, COUNT (value) from test group by key order by count (value), key;
INFO:  evaluating MV: public.test_mv1
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:  mv_rewrite_check_order_clauses_for_mv: ORDER BY clause (key) not found in query
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has fewer sort keys.
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:    creating and adding path for scan on: public.test_mv1
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  no candidate MVs for query involving {public.test_mv1}.
INFO:  evaluating MV: public.test_mv2
INFO:      looking to rewrite GROUP BY but clause list for MV and query differ in length.
INFO:  evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has no ORDER BY.
INFO:        clause (count(value)) can not be supported by MV SELECT list
INFO:  evaluating MV: public.test_union_mv1
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_union_mv2
INFO:      looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  evaluating MV: public.test_mv5
INFO:  mv_rewrite_check_order_clauses_for_mv: looking to rewrite ORDER BY, but MV has fewer sort keys.
INFO:    creating and adding path for scan on: public.test_mv5
INFO:  no candidate MVs for query involving {public.test_mv5}.
INFO:  no candidate MVs for query involving {public.test_mv5}.
                 QUERY PLAN                 
--------------------------------------------
 Sort
   Output: key, (count(value))
   Sort Key: (count(test.value)), test.key
   ->  Custom Scan (MVRewriteScan)
         Output: key, (count(value))
         Rewritten: scan of public.test_mv1
         ->  Seq Scan on public.test_mv1
               Output: key, count
(8 rows)

